common/metadata.go

```diff
@@ -0,0 +1,156 @@
+package common
+
+type UpsertMetadataAction string
+
+const (
+	// UpsertMetadataActionCreate indicates that the object/field was created.
+	UpsertMetadataActionCreate UpsertMetadataAction = "create"
+	// UpsertMetadataActionUpdate indicates that the object/field was updated.
+	UpsertMetadataActionUpdate UpsertMetadataAction = "update"
+	// UpsertMetadataActionNone indicates that the object/field was not changed.
+	UpsertMetadataActionNone UpsertMetadataAction = "none"
+)
+
+// UpsertMetadataParams represents parameters for upserting metadata.
+type UpsertMetadataParams struct {
+	// Maps object names to field definitions.
+	FieldsDefinitions map[string][]*FieldDefinition `json:"customFields"`
+}
+
+// FieldType represents the data type of a field.
+type FieldType string
+
+const (
+	FieldTypeString       FieldType = "string"
+	FieldTypeBoolean      FieldType = "boolean"
+	FieldTypeDate         FieldType = "date"
+	FieldTypeDateTime     FieldType = "datetime"
+	FieldTypeSingleSelect FieldType = "singleSelect"
+	FieldTypeMultiSelect  FieldType = "multiSelect"
+	FieldTypeInt          FieldType = "int"
+	FieldTypeFloat        FieldType = "float"
+)
+
+// IsValid checks if the FieldType is known.
+func (ft FieldType) IsValid() bool {
+	switch ft {
+	case
+		FieldTypeString,
+		FieldTypeBoolean,
+		FieldTypeDate,
+		FieldTypeDateTime,
+		FieldTypeSingleSelect,
+		FieldTypeMultiSelect,
+		FieldTypeInt,
+		FieldTypeFloat:
+		return true
+	default:
+		return false
+	}
+}
+
+// FieldDefinition represents a field definition. Note that not all
+// providers will support all fields. This is a best-effort attempt
+// to create a common schema for custom fields across providers.
+//
+// In the event that a provider doesn't support a particular field,
+// and assuming that the field has a value, then the value should be
+// ignored, and a warning should be added to the UpsertMetadataResult.
+type FieldDefinition struct {
+	// FieldName is the short name of the field, e.g. "My_Custom_Field".
+	FieldName string `json:"fieldName"`
+	// DisplayName is the human-readable name of the field, e.g. "My Custom Field".
+	DisplayName string `json:"displayName"`
+	// Description is an optional description of the field.
+	Description string `json:"description,omitempty"`
+	// ValueType is the data type of the field.
+	ValueType FieldType `json:"valueType"`
+	// Length is the maximum length of the field (for string types).
+	Length *int `json:"length,omitempty"`
+	// Precision is the total number of digits (for decimal types).
```

Comment:
is this the number of digits after the decimal point?

Reply:
From the salesforce docs:

The precision for number values. Precision is the number of digits in a number. For example, the number 256.99 has a precision value of 5.